cpp/src/arrow/dataset/scanner.cc

```diff
@@ -1285,9 +1284,21 @@ Result<compute::ExecNode*> MakeOrderedSinkNode(compute::ExecPlan* plan,
 
   return node;
 }
-compute::ExecFactoryRegistry::AddOnLoad kRegisterSink("ordered_sink",
-                                                      MakeOrderedSinkNode);
 
 }  // namespace
+
+namespace internal {
+
+void Initialize() {
+  static auto registry = compute::default_exec_factory_registry();
+  if (registry) {
+    DCHECK_OK(registry->AddFactory("scan", MakeScanNode));
+    DCHECK_OK(registry->AddFactory("ordered_sink", MakeOrderedSinkNode));
+    DCHECK_OK(registry->AddFactory("augmented_project", MakeAugmentedProjectNode));
+    registry = nullptr;
+  }
+}
```

Comment:
The intent could be more clearly expressed with std::call_once but this works fine.

Reply:
Though, having calls to this scattered all about makes me worry it'll be easy to forget…at least for scanner.cc, maybe the call could be placed in the ctor of AsyncScanner or such so that we don't have to place it in multiple different methods here?

Reply:
The on-load addition is indeed a loss. I'll try a workaround and see if we can keep it